lib/driver-cli.js

```diff
@@ -0,0 +1,265 @@
+import _ from 'lodash';
+import { KNOWN_DRIVERS, validateDriversJson, writeInstalledDriver,
+         INSTALL_TYPE_LOCAL, INSTALL_TYPE_NPM, INSTALL_TYPE_GIT,
+         INSTALL_TYPE_GITHUB } from './drivers';
+import { getLatestVersion, installPackage, linkPackage } from './npm';
+import semver from 'semver';
+
+let INSTALLED_DRIVERS;
+
+function errAndQuit (json, msg) {
+  if (json) {
+    console.log(JSON.stringify({error: msg}));
+  } else {
+    console.error(msg.red);
+  }
+  process.exit(1);
+}
+
+function log (json, msg) {
+  !json && console.log(msg);
+}
+
+async function runDriverCommand (args) {
+  let jsonResult = null;
+  try {
+    INSTALLED_DRIVERS = validateDriversJson();
+    switch (args.driverCommand) {
+      case 'list':
+        jsonResult = await list(args);
+        break;
+      case 'install':
+        jsonResult = await install(args);
+        break;
+      case 'uninstall':
+        jsonResult = await uninstall(args);
+        break;
+      case 'update':
+        jsonResult = await update(args);
+        break;
+      default:
+        throw new Error(`Cannot handle driver command ${args.driverCommand}`);
+    }
+  } catch (err) {
+    errAndQuit(args.json, err.message);
+  }
+
+  if (args.json) {
+    console.log(JSON.stringify(jsonResult));
+  }
+}
+
+function startSpinner (msg) {
+  process.stdout.write(msg);
+
+  let count = 1;
+  const numDots = 5;
+  const dotTime = 500;
+  const cursorStart = msg.length;
+
+  const interval = setInterval(() => {
+    if (count % (numDots + 1) === 0) {
+      process.stdout.cursorTo(cursorStart);
+      for (let i = 0; i < numDots; i++) {
+        process.stdout.write(' ');
+      }
+      process.stdout.cursorTo(cursorStart);
+    } else {
+      process.stdout.write('.');
+    }
+    count++;
+  }, dotTime);
+
+  return () => {
+    clearInterval(interval);
+    process.stdout.write('\n');
+  };
+}
+
+function logSpin (json, msg) {
+  if (json) {
+    return () => {};
+  }
+
+  return startSpinner(msg);
+}
+
+async function list ({json, showInstalled, showUpdates}) {
+  const lsMsg = `Listing ${showInstalled ? 'installed' : 'available'} drivers`;
+  const stopSpinner = logSpin(json, lsMsg);
+
+  const drivers = {};
+
+  for (const name of Object.keys({...INSTALLED_DRIVERS})) {
```

Comment:
what is the purpose of `{...INSTALLED_DRIVERS}` ?

Reply:
I thought I was protecting against mutation of INSTALLED_DRIVERS. but i wasn't.